samples/java_springboot/52.teams-messaging-extensions-search-auth-config/src/main/java/com/microsoft/bot/sample/teamssearchauth/TeamsMessagingExtensionsSearchAuthConfigBot.java

```diff
@@ -44,35 +45,106 @@
  */
 public class TeamsMessagingExtensionsSearchAuthConfigBot extends TeamsActivityHandler {
 
-    private String siteUrl;
-    private String connectionName;
-    private UserState userState;
-    private StatePropertyAccessor<String> userConfigProperty;
+    private final String siteUrl;
+    private final String connectionName;
+    private final UserState userState;
+    private final StatePropertyAccessor<String> userConfigProperty;
 
     public TeamsMessagingExtensionsSearchAuthConfigBot(
         Configuration configuration,
         UserState userState
     ) {
+        if (StringUtils.isBlank(configuration.getProperty("ConnectionName"))) {
+            throw new IllegalArgumentException("ConnectionName cannot be null");
+        }
         connectionName = configuration.getProperty("ConnectionName");
+
+        if (StringUtils.isBlank(configuration.getProperty("SiteUrl"))) {
+            throw new IllegalArgumentException("SiteUrl cannot be null");
+        }
         siteUrl = configuration.getProperty("SiteUrl");
-        userConfigProperty = userState.createProperty("UserConfiguration");
+
+        if (userState == null) {
+            throw new IllegalArgumentException("userState cannot be null");
+        }
         this.userState = userState;
+        userConfigProperty = userState.createProperty("UserConfiguration");
     }
 
     @Override
     public CompletableFuture<Void> onTurn(TurnContext turnContext) {
         return super.onTurn(turnContext)
+            // After the turn is complete, persist any UserState changes.
             .thenCompose(saveResult -> userState.saveChanges(turnContext));
     }
 
+    @Override
+    protected CompletableFuture<MessagingExtensionResponse> onTeamsAppBasedLinkQuery(
+            TurnContext turnContext,
+            AppBasedLinkQuery query) {
+        String state = query.getState(); // Check the state value
+        TokenResponse tokenResponse = this.getTokenResponse(turnContext, state).join();
+        if (tokenResponse == null || StringUtils.isBlank(tokenResponse.getToken())) {
+            // There is no token, so the user has not signed in yet.
+
+            // Retrieve the OAuth Sign in Link to use in the MessagingExtensionResult Suggested Actions
+            UserTokenProvider userTokenClient = (UserTokenProvider) turnContext.getAdapter();
+            SignInResource resource = userTokenClient.getSignInResource(turnContext, this.connectionName).join();
```

Comment:
Why we are adding these `join()` calls?

Reply:
As we discussed, I replaced the `thenApply` and `thenCompose` to `join()` in order to be sync. I think that the best approach is to have the thenApply for sure, but all the scenarios in teams are different..

I want to keep it as is